test/readme.md

```diff
@@ -1,110 +1,268 @@
+
+1. Start a sharded cluster. You can use the [cluster_setup.sh](tools/cluster_setup.sh) script to do so: `./test/tools/cluster_setup.sh sharded_cluster`. The tool should create a cluster with two mongos, so you have a URI similar to `MONGODB_URI=mongodb://host1,host2/`.
+1. Create an environment variable named `MONGODB_URI` that stores the URI of the sharded cluster you just created. For example: `export MONGODB_URI="mongodb://host1,host2/"`
+1. Install the HAProxy load balancer. For those on macOS, you can install HAProxy with `brew install haproxy`.
+1. Start the load balancer by using the [run-load-balancer script](https://github.com/mongodb-labs/drivers-evergreen-tools/blob/master/.evergreen/run-load-balancer.sh) provided in drivers-evergreen-tools.
+   ```sh
+   $DRIVERS_TOOLS/.evergreen/run-load-balancer.sh start
+   ```
+   A new file name `lb-expansion.yml` will be automatically created. The contents of the file will be similar in structure to the code below.
+   ```yaml
+   SINGLE_MONGOS_LB_URI: "mongodb://127.0.0.1:8000/?loadBalanced=true"
+   MULTI_MONGOS_LB_URI: "mongodb://127.0.0.1:8001/?loadBalanced=true"
+   ```
+1. Generate a sourceable environment file from `lb-expansion.yml` by running the following command:
+   ```sh
+   cat lb-expansion.yml | sed 's/: /=/g' > lb.env
+   ```
+   A new file name `lb.env` is automatically created.
+1. Add an additional environment variable named `FAKE_MONGODB_SERVICE_ID` to the end of the `lb.env` file. Setting `FAKE_MONGODB_SERVICE_ID` to `true` enables logic in the driver to stick in a fake service ID on responses since that's what a real load balanced deployment is required to do.
+   ```sh
+   FAKE_MONGODB_SERVICE_ID="true"
+   ```
+   > **Note:** `FAKE_MONGODB_SERVICE_ID` will no longer be needed with the completion of [NODE-3431](https://jira.mongodb.org/browse/NODE-3431).
+1. Source the environment variables using a command like `source lb.env`.
+1. Export **each** of the environment variables that were created in `lb.env`. For example: `export SINGLE_MONGOS_LB_URI`.
+1. Run the test suite as you normally would:
+   ```sh
+   npm run check:test
+   ```
+   Verify that the output from Mocha includes `[ topology type: load-balanced ]`. This indicates the tests successfully accessed the specialized environment variables for load balancer testing.
+1. When you are done testing, shutdown the HAProxy load balancer:
+   ```sh
+   $DRIVERS_TOOLS/.evergreen/run-load-balancer.sh stop
+   ```
+
+### Client-Side Field Level Encryption (CSFLE)
+
+The following steps will walk you through how to run the tests for CSFLE.
+
+1. Install [MongoDB Client Encryption][npm-csfle] if you haven't already:
+   `npm install mongodb-client-encryption`
+
+1. Create the following environment variables using a command like `export AWS_REGION="us-east-1"`.
+
+   - `AWS_ACCESS_KEY_ID`
+   - `AWS_SECRET_ACCESS_KEY`
+   - `CSFLE_KMS_PROVIDERS`
```

Comment:
I'd describe CSFLE_KMS_PROVIDERS as "The raw EJSON description of the KMS providers". The format is:

```
interface CSFLE_kms_providers {
   aws: {
     accessKeyId: string;
     secretAccessKey: string;
  };
  azure: {
    tenantId: string;
    clientId: string;
    clientSecret: string;
  };
  gcp: {
    email: string;
    privateKey: string;
  };
  local: {
    // EJSON handle converting this, its actually the canonical -> { $binary: { base64: string; subType: string } }
    // **NOTE**: The dollar sign has to be escaped when using this as an ENV variable
    key: Binary;
  }
}
```

Reply:
Than you for this!